### PR TITLE
fix: add state root check while syncing

### PR DESCRIPTION
**Pull request by vaibhavchellani**
Fixes pubkey serialisation and API related bugs

### CODE REVIEW

**Comment by vaibhavchellani** [comment]
There will be a follow up PR with aggregated signature check post this one.

File: listener/processors.go [hunk 1]
```diff
@@ -185,6 +189,10 @@ func (s *Syncer) processNewBatch(eventName string, abiObject *abi.ABI, vLog *eth
 	if err != nil && gorm.IsRecordNotFoundError(err) {
 		s.Logger.Info("Found a new batch, applying transactions and adding new batch", "index", event.Index.Uint64)
 		_, commitments, err := s.parseAndApplyBatch(vLog.TxHash, event.BatchType)
+		if err == ErrBadBatch {
+			s.Logger.Error("Fraud detected, please challenge batch on-chain", "error", err)
+			panic(err)
```
Comment (ChihChengLiang) [inline listener/processors.go:194]: Just highlight that this is the current way we handle the fraud.
Nothing to change here. But we might want to think about the flow to challenge automatically.

File: rest/handler.go [hunk 2]
```diff
@@ -58,37 +59,51 @@ func TxHandler(w http.ResponseWriter, r *http.Request) {
 	}
 	txMessageBytes, err := hex.DecodeString(tx.Message)
 	if err != nil {
-		WriteErrorResponse(w, http.StatusBadRequest, "Cannot decode message")
+		WriteErrorResponse(w, http.StatusBadRequest, err.Error())
```
Comment (ChihChengLiang) [inline rest/handler.go:62]: Sorry I was not saying clearly in the call. But let me send another pr to illustrate the idea.

File: core/pubkey.go [hunk 3]
```diff
@@ -20,30 +20,30 @@ const pubkeyLength = 128
...
 func NewPubkey(p [4]*big.Int) Pubkey {
 	pubkey := make([]byte, pubkeyLength)
-	copy(pubkey[:32], common.LeftPadBytes(p[1].Bytes(), 32))
-	copy(pubkey[32:64], common.LeftPadBytes(p[0].Bytes(), 32))
-	copy(pubkey[64:96], common.LeftPadBytes(p[3].Bytes(), 32))
-	copy(pubkey[96:128], common.LeftPadBytes(p[2].Bytes(), 32))
+	copy(pubkey[:32], common.LeftPadBytes(p[0].Bytes(), 32))
+	copy(pubkey[32:64], common.LeftPadBytes(p[1].Bytes(), 32))
+	copy(pubkey[64:96], common.LeftPadBytes(p[2].Bytes(), 32))
+	copy(pubkey[96:128], common.LeftPadBytes(p[3].Bytes(), 32))
```
Comment (ChihChengLiang) [inline core/pubkey.go:26]: Nice fixing the public keys here.

File: cmd/main.go [hunk 4]
```diff
@@ -96,10 +96,14 @@ func createUsers() *cobra.Command {
...
 				return err
 			}
+			appID, err := core.HexToByteArray(cfg.AppID)
+			if err != nil {
+				return err
+			}
 
 			var users []User
 			for i := 0; i < userCount; i++ {
-				newWallet, err := wallet.NewWallet(core.StringToByteArray(cfg.AppID))
```
Comment (ChihChengLiang) [inline cmd/main.go:102]: Eew ... So StringToByteArray treats the input string as ascii or utf8 encoded, while HexToByteArray treats it as hex, which is what we want.

File: rest/handler.go [hunk 5]
```diff
@@ -484,7 +501,7 @@ func estimateNonceHandler(w http.ResponseWriter, r *http.Request) {
 
 	var resp estimateNonceResp
 	resp.StateID = uint64(idInt)
-	resp.Nonce = pendingNonce
+	resp.Nonce = pendingNonce + 1
```
Comment (ChihChengLiang) [inline rest/handler.go:504]: This looks strange to me. This line suggests the first transaction is nonce 1. But in the contract, the first transaction is nonce 0.
But we can unblock this pr, and talk about this more later.

**Review by ChihChengLiang — APPROVED** [review]
Yay, fix much stuff. Left some comments here. Most of them for future improvement, nothing need to change now. Looks good to me.

File: rest/handler.go (hunk 2 above)
Comment (ChihChengLiang) [reply rest/handler.go:62]: #148 is the pr.

Comment (vaibhavchellani) [reply rest/handler.go:62]: Ah I see what you mean, but what do you think of retaining control of what error code we want to send over an error, the current implementation was using `BadRequest` everywhere and there was a fix coming for custom status code with error since that is required by wallets and explorers to detect what went wrong.

File: listener/processors.go (hunk 1 above)
Comment (vaibhavchellani) [reply listener/processors.go:194]: Yeah there will be a way to autochallenge and plan to do that later, let me highlight that here

File: rest/handler.go (hunk 5 above)
Comment (vaibhavchellani) [reply rest/handler.go:504]: Sounds good

File: cmd/main.go (hunk 4 above)
Comment (vaibhavchellani) [reply cmd/main.go:102]: yeah, no wonder I wasn't using this, it was a tech debt from a previous version, thankfully it's no more.

File: rest/handler.go (hunk 2 above)
Comment (ChihChengLiang) [reply rest/handler.go:62]: Alright, if that's the case, the current approach looks fine.
